File: 1909-buildings-with-an-ocean-view/buildings-with-an-ocean-view.java
class Solution {
    public int[] findBuildings(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.addFirst(heights.length - 1);
        for(int i = heights.length - 2; i >= 0; i--) {
            if(heights[i] > heights[stack.peekFirst()]) stack.addFirst(i);
        }
        
        int[] result = new int[stack.size()];
        for(int i = 0; i < result.length; i++) result[i] = stack.removeFirst();
        return result;
    }
}

File: 1909-buildings-with-an-ocean-view/buildings-with-an-ocean-view.cpp
class Solution {
public:
    vector<int> findBuildings(vector<int>& heights) {
        stack<int> stack;
        vector<int> result;
        stack.push(heights.size() - 1);
        for(int i = heights.size() - 2; i >= 0; i--){
            if(heights[i] > heights[stack.top()]){
                stack.push(i);
            }
        }
        while(!stack.empty()){
            result.push_back(stack.top());
            stack.pop();
        }
        return result;
    }
};